lib/grpc/stream.ex

```diff
@@ -1,17 +1,433 @@
+
+  ## Caution
+
+  This version propagates errors via raised exceptions, which can crash the Flow worker and halt the pipeline.
+  Prefer `ask/3` for production use unless failure should abort the stream.
+  """
+  @spec ask!(t(), pid | atom, non_neg_integer) :: t()
+  defdelegate ask!(stream, target, timeout \\ 5000), to: Operators
+
+  @doc """
+  Replaces the stream with a new one generated by the given factory function.
+  The factory function receives the metadata and the current item as arguments.
+  The factory function should return a new `GRPC.Stream` struct.
+
+  ## Caution
+  This function materializes the current stream and replaces it with a new one.
+  This can lead to performance issues if used excessively. In fact, it is more commonly used
+  in following the `single/2` of `single_as_ctx/3` functions. 
+
+  ## Parameters
+
+    - `stream`: The current `GRPC.Stream`.
+    - `factory`: A function that takes metadata and the current item, returning a new `GRPC.Stream`.
+
+  ## Returns
+    - A new `GRPC.Stream` struct with the replaced flow.
+
+  ## Example
+
+    GRPC.Stream.single(request)
+    # Create new output stream from the request and drop the input
+    |> GRPC.Stream.replace(fn _metadata, msg ->
+      create_output_stream(msg) # returns some Enumerable
+      |> GRPC.Stream.from()
+    end)
+    |> GRPC.Stream.run_with(materializer)
+  """
+  @spec replace(t(), (map(), term -> term)) :: t()
+  defdelegate replace(stream, factory), to: Operators
+
+  @doc """
+  Filters the stream using the given predicate function in parallel.
+  """
+  @spec filter(t(), (term -> term)) :: t
+  defdelegate filter(stream, filter), to: Operators
+
+  @doc """
+  Applies a mapping function and flattens the results (one level deep).
+
+  Useful for emitting multiple messages for each input.
+  """
+  @spec flat_map(t, (term -> Enumerable.t())) :: t()
+  defdelegate flat_map(stream, flat_mapper), to: Operators
+
+  @doc """
+  Applies a transformation function to each stream item.
+  """
+  @spec map(t(), (term -> term)) :: t()
+  defdelegate map(stream, mapper), to: Operators
+
+  @doc """
+  Applies a transformation function to each stream item, passing the context as an additional argument.
+  This is useful for operations that require access to the stream's headers.
+  """
+  @spec map_with_ctx(t(), (map(), term -> term)) :: t()
+  defdelegate map_with_ctx(stream, mapper), to: Operators
+
+  @doc """
+  Partitions the stream to allow grouping of items by key or condition.
+
+  Use this before stateful operations such as `reduce/3`.
+
+  ## Note
+
+  Excessive use of partitioning can impact performance and memory usage.
+  Only partition when required for correctness or performance.
+  """
+  @spec partition(t(), keyword()) :: t()
+  defdelegate partition(stream, options \\ []), to: Operators
+
+  @doc """
+  Reduces items in the stream using an accumulator.
+
+  ## Parameters
+
+    - `acc_fun` initializes the accumulator,
```

Comment:
why do we need a 0-arity function for acc? Is this just Flow leaking?

Reply:
It was mapped 1:1 from `Flow` api but this is perfectly normal for any flow engine. This exists in Flow, in Akka, in Python, in Rust frameworks. I see no harm in supporting it.

Reply:
I think for all functions that have this kind of Flow abstraction "leaking" we should have something like: see Flow.<function> for more information on <argument>

Reply:
Just lean onto Flow directly

Reply:
Docs updates!